k2/csrc/array.h

```diff
@@ -103,19 +160,49 @@ struct Array3 {
                            // .. indexes2[indexes1[size]-1] should be defined.
 
   Ptr data;  // `data` might be an actual pointer, or might be some object
-             // supporting operator [].  data[indexes[0]] through
-             // data[indexes[size] - 1] must be accessible through this
+             // supporting operator [].  data[indexes2[indexes1[0]]] through
```

Comment:
Actually in current implementation, we only accept `Ptr` as a pointer as we compare it with `null_ptr` in `Array1` and `Array2`, we don't support type such as `vector<valueType>`, so we should update the documentation accordingly. But I guess @danpovey may want to support those type (supporting `[ ]` but not a pointer) when writing the inferface.

Reply:
BTW, name `Ptr` is a not good choice if we want to support type such as `vector`, we also may need to write traits to get value type from non-pointer type  that supporting operator `[]`

Reply:
Mm, OK.  What I actually had in mind here was StridedPtr, which is like regular pointer with a stride.  We could support comparison with nullptr there.  I don't think 'vector' is necessary.  You may be right about the traits, or we could rely on auto.

Reply:
We need value type as the return type, so we cannot rely on auto. We could just follow STL conventions which define `value_type` in `StridePtr`, then we can use `typename std::iterator_traits<T>::value_type` to get the value type, as `std::iterator_traits` has alreadly defined specialized version for `T*` and `const T*`, or we can define our own traints if you think name `iterator_traits` is not good here.

Reply:
I am wondering how `StridePtr` would be used.

- For 1-d arrays, there is one stride.
- For 2-d arrays, there are two strides, one for the columns and another for the rows.
- For n-d arrays, there are N strides.

How can we set the `stride` in the `StridePtr` for different types of arrays?

Reply:
I guess there is only 1-d situation for StridePtr in current infterface, for `n >= 2`, we actually control stride with those `indexes` (indexes1, indexes2,...)